### PR TITLE
Adds in Virtual Page view for tracking filters.

### PR DESCRIPTION
**Pull request by DilwoarH**
User Story:
As a delivery team I need to understand which filters have been selected so that I can make decisions about which filters to improve or remove.

Virtual Page View class has been refactored to follow JS standards as well as including a new trackFilterAnalytics function.

Ticket: https://trello.com/c/o355GRIi/513-tracking-filters

### CODE REVIEW

**Comment by DilwoarH** [comment]
Required: https://github.com/alphagov/digitalmarketplace-frontend-toolkit/pull/362

**Comment by DilwoarH** [comment]
Need to write tests

File: app/assets/javascripts/analytics/_virtualPageViews.js [hunk 1]
```diff
@@ -1,19 +1,35 @@
...
+      url         = urlList.join("?");
       GOVUK.analytics.trackPageview(url);
     }
+
   };
 
-  GOVUK.GDM.analytics.virtualPageViews = function() {
-    $('[data-analytics=trackPageView]').each(sendVirtualPageView);
+  VirtualPageViews.prototype.trackFilterAnalytics = function( key, value ) {
+    var url = '/filters/' + key + '/' + value;
```
Comment (DilwoarH) [inline app/assets/javascripts/analytics/_virtualPageViews.js]: Needs to be changed to `/g-cloud/filters/{group}/{key}/{value}`

File: app/assets/javascripts/analytics/_virtualPageViews.js [hunk 2]
```diff
@@ -1,19 +1,35 @@
...
+      urlList[0]  = urlList[0] + "/vpv";
+      url         = urlList.join("?");
       GOVUK.analytics.trackPageview(url);
     }
+
   };
 
-  GOVUK.GDM.analytics.virtualPageViews = function() {
-    $('[data-analytics=trackPageView]').each(sendVirtualPageView);
+  VirtualPageViews.prototype.trackFilterAnalytics = function( key, value ) {
```
Comment (DilwoarH) [inline app/assets/javascripts/analytics/_virtualPageViews.js]: needs to allow group to be passed in

**Comment by DilwoarH** [comment]
Tests updated as well as new tests added for filter page views

File: spec/javascripts/unit/AnalyticsSpec.js [hunk 3]
```diff
@@ -187,32 +187,74 @@ describe("GOVUK.Analytics", function () {
...
+            '</div>' +
+          '</form>'
+      );
+      $(document.body).append($analyticsString);
+      window.GOVUK.GDM.analytics.virtualPageViews.init();
+      $('#filter-option-1').click();
+      expect(window.ga.calls.first().args[2]).toEqual({page: "/g-cloud/filters/example-filters-group/filter-option-1/filter-option-1-value/vpv"});
+    });
+
+    it("Should not trigger virtual page view on filter selection", function () {
```
Comment (galund) [inline spec/javascripts/unit/AnalyticsSpec.js]: I understood the first test, but this one seems identical (except that we don't expect it to trigger vpv, and it's not immediately obvious why?)

File: tests/main/presenters/test_search_presenters.py [hunk 4]
```diff
@@ -86,6 +86,7 @@ def test_get_filter_groups_from_questions_with_radio_filters(self):
 
         assert radios_filter_group == {
             'label': 'Radios example',
+            'slug': 'radios-example',
```
Comment (galund) [inline tests/main/presenters/test_search_presenters.py]: Out of interest, what happens if these aren't there?

File: spec/javascripts/unit/AnalyticsSpec.js (hunk 3 above)
Comment (DilwoarH) [inline spec/javascripts/unit/AnalyticsSpec.js]:
```    it("Should not trigger virtual page view if a user is removing a filter", function () {
```

Is this a better title?

File: tests/main/presenters/test_search_presenters.py (hunk 4 above)
Comment (DilwoarH) [inline tests/main/presenters/test_search_presenters.py]: test fails if slug does not exist in the assertion.

**Comment by DilwoarH** [comment]
@galund - Added additional requirements by Ash. Titles are now being sent to Google Analytics in the following format:  `Filter - {group} - {filter_label} - {filter_value}`;

File: app/assets/javascripts/analytics/_virtualPageViews.js [hunk 5]
```diff
@@ -1,19 +1,53 @@
...
+
   };
 
-  GOVUK.GDM.analytics.virtualPageViews = function() {
-    $('[data-analytics=trackPageView]').each(sendVirtualPageView);
+  VirtualPageViews.prototype.trackFilterAnalytics = function( group, key, value ) {
+
+    var filter_path = '/' + group + '/' + key + '/' + value;
+    var url = '/g-cloud/filters' + filter_path;
+    this.sendVirtualPageView( url, 'Filter' + filter_path.replace(/[/]/g, " - ") );
```
Comment (galund) [inline app/assets/javascripts/analytics/_virtualPageViews.js]: I suppose we could find a way to pass in the actual human-readable value, for the page title, but we can come back to that if it turned out to be helpful :)

File: app/assets/javascripts/analytics/_virtualPageViews.js [hunk 6]
```diff
@@ -1,19 +1,53 @@
...
-      GOVUK.analytics.trackPageview(url);
+  var VirtualPageViews = function(){};
+
+  VirtualPageViews.prototype.init = function() {
+
+    var _self = this;
+
+    $('[data-analytics=trackPageView]').each(function(){
+      var $element = $(this);
+      var url = $element.data('url');
```
Comment (DilwoarH) [inline app/assets/javascripts/analytics/_virtualPageViews.js]: This was copied from the original code but good point, I'll update it.

File: app/assets/javascripts/analytics/_virtualPageViews.js [hunk 7]
```diff
@@ -1,19 +1,53 @@
...
-    if (GOVUK.analytics  && url){
-      var urlList = url.split("?")
-      urlList[0] = urlList[0] + "/vpv"
-      url = urlList.join("?")
-      GOVUK.analytics.trackPageview(url);
+  var VirtualPageViews = function(){};
+
+  VirtualPageViews.prototype.init = function() {
+
+    var _self = this;
```
Comment (DilwoarH) [inline app/assets/javascripts/analytics/_virtualPageViews.js]: change to `_this`

File: app/assets/javascripts/analytics/_virtualPageViews.js (hunk 5 above)
Comment (DilwoarH) [inline app/assets/javascripts/analytics/_virtualPageViews.js]: yeah would be good

**Pull request merged**